# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/LeylaAghayeva1/ml-search-engineering/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Selected Lane: Refresh / Content Opportunity Scoring

I chose the Refresh / Content Opportunity Scoring lane because it focuses on helping editors decide which pages should be reviewed first when they have limited time and resources. Instead of trying to update every page, the goal is to rank pages according to their potential need for review. This matches the objective of the starter project, which is to produce a ranked review queue rather than simply predict whether a page is declining. It also aligns with the evaluation used in the starter pipeline (Precision@50), making it a suitable foundation for the next seven weeks.

In [14]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

df["is_declining_label"] = (df["trend_direction"] == "down").astype(int)
print(f"Rows: {len(df):,}")
print(f"Declining pages: {df['is_declining_label'].sum():,}")
print(f"Declining rate: {df['is_declining_label'].mean():.3f}")

Rows: 30,000
Declining pages: 16,262
Declining rate: 0.542


## 2. The question: decision, action, cost of a wrong call

### Research question

**Which content pages should editors review first to reduce the risk of losing search visibility?**

### Decision

Determine which pages should be prioritized for content review and refresh.

### Who acts?

Website editors or content managers.

### Action

Editors review the highest-ranked pages and decide whether to refresh, expand, rewrite, merge, or monitor the content.

### Cost of a wrong recommendation

A **false positive** means an editor spends time reviewing a page that did not actually need attention.

A **false negative** means an important declining page is missed, which may lead to continued loss of search traffic and missed business opportunities.

### Why this is not just "train a model"

The goal is not simply to predict which pages are declining. The goal is to improve a business decision by helping editors prioritize their limited time. The model is only a tool for producing a better review queue.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

# Load the starter dataset
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

print(f"Dataset shape: {df.shape}")
print(f"Unique clients: {df['client_id'].nunique()}")
print(f"Unique content pages: {df['content_id'].nunique()}")

Dataset shape: (30000, 44)
Unique clients: 32
Unique content pages: 30000


## 3. Quick look at the data (2–3 real numbers)

The starter dataset contains **30,000** content pages collected from **32** clients.

Among these pages, **16,262** are labeled as declining, representing approximately **54.2%** of the dataset.

These numbers suggest that content decline is common enough to make prioritization valuable. Since reviewing all 30,000 pages manually would be impractical, a ranking system can help editors focus on the pages most likely to benefit from review.

In [15]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import pandas as pd

# Load the starter dataset
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")
# Create the target label
df["is_declining_label"] = (df["trend_direction"] == "down").astype(int)
# Calculate supporting statistics
total_pages = len(df)
total_clients = df["client_id"].nunique()
declining_pages = df["is_declining_label"].sum()
declining_rate = df["is_declining_label"].mean() * 100
print(f"Total content pages: {total_pages:,}")
print(f"Unique clients: {total_clients:,}")
print(f"Declining pages: {declining_pages:,}")
print(f"Declining rate: {declining_rate:.1f}%")

Total content pages: 30,000
Unique clients: 32
Declining pages: 16,262
Declining rate: 54.2%


## 4. Careful words: what I can and can't claim

This project can make **observed and directional claims** based on the patterns found in the available content and search data.

The work can identify pages that show signals associated with lower performance or higher refresh priority. It can provide a ranked list that helps editors decide which pages to review first.

However, this project cannot claim that refreshing a page will definitely improve traffic. It cannot prove that one factor caused a page to decline because the data is observational rather than experimental.

This project also cannot claim to predict Google's algorithm or explain ranking changes in search engines. The output should be treated as **decision support for human editors**, not an automated replacement for editorial judgment.

In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Check available columns for product decision fields
product_fields = [
    "health_score",
    "priority_score",
    "action_type"
]

available_product_fields = [
    col for col in product_fields if col in df.columns
]

print("Product decision fields found:")
print(available_product_fields)

if len(available_product_fields) == 0:
    print("No product decision fields found. Dataset contains observable signals only.")

Product decision fields found:
[]
No product decision fields found. Dataset contains observable signals only.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.